### PR TITLE
Signatures

### PR DESCRIPTION
**Pull request by conartist6**
I think the changes to API.md should be more or less self-explanatory.

### CODE REVIEW

**Comment by KSXGitHub** [comment]
Why don't you just use jsdoc or typedoc to generate a website?

**Comment by conartist6** [comment]
I cribbed the square brace thing from man pages, but with command line commands square braces don't have such an obvious conflict of meaning. Perhaps a different marker would be more appropriate, like angle braces or pipes:
```
slice(<start, <end, <step>>>, source)
slice(|start, |end, |step|||, source)
```

**Comment by conartist6** [comment]
I think I like angle braces.

**Comment by conartist6** [comment]
@KSXGitHub just noticed your comment. I still plan to make a website. I don't want to use anything specific to a particular type language, because I think of the library as js first, ts second.

More than anything else though I just don't think that either jsdoc or esdoc will afford me the flexibility I want, especially since I want to build the docs from the sources, not just from comments which need to be kept in sync with the sources.

And I do still plan to build a website. I don't know if there will be any point in keeping the `API.md` file once there is a website, but certainly most of the work in this PR would also be used in the data model for a website.

**Comment by conartist6** [comment]
Even the things I just discussed in terms of how optional arguments need to be communicated is beyond the abilities of any docgen I've ever seen.

**Comment by KSXGitHub** [comment]
> Even the things I just discussed in terms of how optional arguments need to be communicated is beyond the abilities of any [off the shelf] docgen I've ever seen.

Because there're optional parameters that predate obligated parameters and auto currying?

**Comment by KSXGitHub** [comment]
Anyway, TypeScript have overloaded functions, which is clearer and more explicit than the compacted signature you invented.

**Comment by conartist6** [comment]
Because our optional arguments usually go in front of other arguments, shifting the positions of other positional arguments around. To a large extent we can't avoid this since the position of the iterable arguments must be able to shift in order to allow non-curried calls.

I would also be open to expressing this as overloads. Most methods only have one optional parameter at most, which would equate to two overloads. At most some functions would have three overloads. Our bread and butter functions like `map` and `filter` would have only one overload, which is good.

**Comment by conartist6** [comment]
@KSXGitHub I've updated the PR. Take a look at [range](https://github.com/iter-tools/iter-tools/blob/a9844f3c99e0ded48e8c542789cd1bdbd1b2c9ab/API.md#range) for example.

**Comment by conartist6** [comment]
I think having the autogenerated signatures section frees me up to include fewer (but more focused) examples. I don't need the examples to show the results of equivalent signatures anymore, just the different cases of runtime behavior.

File: API.md [hunk 1]
```diff
@@ -156,6 +197,8 @@ range(3, 10, 3); // 3, 6, 9
 ```
 
 ### repeat
+**repeat(item, nTimes)**
```
[the remark this answers is not in the document]
Comment (KSXGitHub) [reply API.md]: You can instead write:

```
repeat(item, nTimes?)
```

[the remark this answers is not in the document]
Comment (conartist6) [reply API.md]: I can, yes.

**Comment by conartist6** [comment]
Looks like `splitOn` and `splitWith` methods didn't get signatures generated for them.

**Comment by conartist6** [comment]
OK I think I'm done messing with this now. I've gone through `API.md` and every single method has a signature or signatures now. I'm sure some final tweaks will continue to filter in, but I think I've achieved what I set out to. Thanks @KSXGitHub for pushing me towards the overloads solution; it has worked out nicely.

**Pull request merged**